### ISSUE TITLE
Pyright ignores Any annotations in recursive dict

### ISSUE THREAD

**Issue by kumaraditya303** [microsoft/pyright#2342]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Pyright ignore Any annotations in recursive dict

**To Reproduce**

```py
from typing import Any, Dict


def func(d: Dict[str, Any]) -> None:
    d["hello"] = {"hello1": "world"}
    d["hello"]["hello1"] = True

```

**Expected behavior**
No error, even mypy does not error.

**VS Code extension or command-line**

```console
pyright 1.1.171
```

**Comment by erictraut** [comment]
This is expected behavior. Pyright performs type narrowing upon assignment of subscripted expressions. Mypy doesn't support this feature, which is why it doesn't emit an error in this case.

The same error will occur for simple variable assignments:
```python
x = {"hello1": "world"}
x["hello1"] = True  # Error
```

You can eliminate the diagnostic through the use of a temporary variable or a cast.
```python
def func(d: Dict[str, Any]) -> None:
    x: Dict[str, Any] = {"hello1": "world"}
    d["hello"] = x
    d["hello"]["hello1"] = True
```

or

```python
def func(d: Dict[str, Any]) -> None:
    d["hello"] = cast(Dict[str, Any], {"hello1": "world"})
    d["hello"]["hello1"] = True
```

**Comment by kumaraditya303** [comment]
Since the values of the dict is explicitly `Any` then pyright should not check it IMO, it should work in assignment but not in cases where Any is given.

**Comment by kumaraditya303** [comment]
> This is expected behavior. Pyright performs type narrowing upon assignment of subscripted expressions. Mypy doesn't support this feature, which is why it doesn't emit an error in this case.
> 
> The same error will occur for simple variable assignments:
> 
> ```python
> x = {"hello1": "world"}
> x["hello1"] = True  # Error
> ```
> 
> You can eliminate the diagnostic through the use of a temporary variable or a cast.
> 
> ```python
> def func(d: Dict[str, Any]) -> None:
>     x: Dict[str, Any] = {"hello1": "world"}
>     d["hello"] = x
>     d["hello"]["hello1"] = True
> ```
> 
> or
> 
> ```python
> def func(d: Dict[str, Any]) -> None:
>     d["hello"] = cast(Dict[str, Any], {"hello1": "world"})
>     d["hello"]["hello1"] = True
> ```

Both of your provided solutions are inefficient because in the first there are unnecessary locals lookups and in latter is a redundant cast which is what I want to avoid

**Comment by erictraut** [comment]
There are two dict values involved in your code. The first has a declared type of `Dict[str, Any]`. The second dict, which is a value stored within the first dict, has an inferred type of `Dict[str, str]`. You are attempting to assign a value to the second dict, and that's what is resulting in the diagnostic. As you point out, the first dict has a declared value type of `Any`, and pyright allows you to assign a value of any type to it.

```python
def func(d: Dict[str, Any]) -> None:
    d["hello"] = {"hello1": "world"}
    d["hello"] = True   # This does not generate an error
```

**Comment by kumaraditya303** [comment]
@erictraut This is what I think should happen here :
```py
from typing import *


def func(d: Dict[str, Any]) -> None:
    d["hello"] = {"hello1": "world"}
    reveal_type(d) # this is "Dict[str, Any]"
    reveal_type(d["hello"]) # this is "dict[str, str]" whereas it should be Any decalred above 

```

**Comment by erictraut** [comment]
I understand your viewpoint, but making this change would eliminate some valuable functionality in pyright, and I'm not convinced that's the right tradeoff. Consider the following example.

```python
1 + "a"  # Generates an error, as you would expect

d: Dict[str, Any] = {}
d["x"] = 1
d["y"] = "a"

d["x"] + d["y"]  # Will not generate error with your proposed change
```

Type narrowing on assignment for subscripts was a feature that had been requested by numerous users and has been in place for the past year. It's commonly used with dictionaries, lists and sets that have `Any` value types. Making the change that you're suggesting would effectively eliminate this feature.

We can revisit this tradeoff in the future if we get enough additional feedback. So far, you are the first and only person who has requested a change here.

**Comment by kumaraditya303** [comment]
> I understand your viewpoint, but making this change would eliminate some valuable functionality in pyright, and I'm not convinced that's the right tradeoff. Consider the following example.
> 
> ```python
> 1 + "a"  # Generates an error, as you would expect
> 
> d: Dict[str, Any] = {}
> d["x"] = 1
> d["y"] = "a"
> 
> d["x"] + d["y"]  # Will not generate error with your proposed change
> ```
> 
> Type narrowing on assignment for subscripts was a feature that had been requested by numerous users and has been in place for the past year. It's commonly used with dictionaries, lists and sets that have `Any` value types. Making the change that you're suggesting would effectively eliminate this feature.
> 
> We can revisit this tradeoff in the future if we get enough additional feedback. So far, you are the first and only person who has requested a change here.

Hi, Thanks for your replies,

Btw, I am trying to use pyright for https://github.com/microsoft/playwright-python which currently mostly type annotated and works with mypy just fine. I want to use pyright because it contains better handling for some of the things and implements more features than mypy. 

I understand your concern but for projects which want to use both pyright and mypy and if pyright would require such changes which are basically declaring temp vars and redundant casts, it becomes hard to justify it. Is there a cli flag to enable such functionality which I am looking for ?

**Comment by erictraut** [comment]
There are many differences between the behavior of mypy and pyright. It has never been our goal to be 100% compatible. Where it makes sense, we try to adopt compatible behavior, but there are many places where it intentionally differs. This is what allows pyright to "contain better handling for some of the things and implement more features".

If you want to use both mypy and pyright on the same project, you will need to make some accommodations. At this time, we have no plans to add a compatibility mode to mirror all of mypy's behaviors.
